### ISSUE TITLE
Errors and warnings on ubuntu/gcc/make

### ISSUE THREAD

**Issue by theKidOfArcrania** [josh-127/Combust#10]
Okay I get a bunch of errors... dunno how you are dealing with it:
```
g++ -g -std=gnu++14 -Wall -Wextra -ocompile main.cc lexer.cc logger.cc source.cc
lexer.cc: In member function ‘Rc<SyntaxToken> Lexer::ReadTokenDirect()’:
lexer.cc:38:14: warning: decomposition declaration only available with -std=c++1z or -std=gnu++1z
         auto [token, asStrayToken, isCommentToken, isEofToken] = ReadTokenOnce();
              ^
lexer.cc:43:50: error: taking address of temporary [-fpermissive]
                 &l->CurrentToken->GetLexemeRange(),
                                                  ^
lexer.cc: In member function ‘std::tuple<char, int, int> Lexer::DecodeNewLineEscape()’:
lexer.cc:92:10: warning: decomposition declaration only available with -std=c++1z or -std=gnu++1z
     auto [firstChar, firstCharLength] = DecodeTrigraph();
          ^
lexer.cc: In member function ‘void Lexer::IncrementCursor()’:
lexer.cc:126:10: warning: decomposition declaration only available with -std=c++1z or -std=gnu++1z
     auto [charValue, charLength, trailingWhitespaceLength] = DecodeNewLineEscape();
          ^
lexer.cc: In member function ‘void Lexer::GetTokenRange(Rc<SyntaxToken>, PSOURCE_RANGE)’:
lexer.cc:166:17: warning: unused variable ‘base’ [-Wunused-variable]
     const char *base{ &l->Source->Lines[line][column] };
                 ^~~~
lexer.cc: In member function ‘void Lexer::SkipIntSuffixes(Rc<SyntaxToken>)’:
lexer.cc:265:84: error: no matching function for call to ‘transform(std::__cxx11::basic_string<char>::iterator, std::__cxx11::basic_string<char>::iterator, std::__cxx11::basic_string<char>::iterator, <unresolved overloaded function type>)’
     std::transform(ciSuffix.begin(), ciSuffix.end(), ciSuffix.begin(), std::toupper);
                                                                                    ^
In file included from /usr/include/c++/7/algorithm:62:0,
                 from lexer.cc:9:
/usr/include/c++/7/bits/stl_algo.h:4295:5: note: candidate: template<class _IIter, class _OIter, class _UnaryOperation> _OIter std::transform(_IIter, _IIter, _OIter, _UnaryOperation)
     transform(_InputIterator __first, _InputIterator __last,
     ^~~~~~~~~
/usr/include/c++/7/bits/stl_algo.h:4295:5: note:   template argument deduction/substitution failed:
lexer.cc:265:84: note:   couldn't deduce template parameter ‘_UnaryOperation’
     std::transform(ciSuffix.begin(), ciSuffix.end(), ciSuffix.begin(), std::toupper);
                                                                                    ^
In file included from /usr/include/c++/7/algorithm:62:0,
                 from lexer.cc:9:
/usr/include/c++/7/bits/stl_algo.h:4332:5: note: candidate: template<class _IIter1, class _IIter2, class _OIter, class _BinaryOperation> _OIter std::transform(_IIter1, _IIter1, _IIter2, _OIter, _BinaryOperation)
     transform(_InputIterator1 __first1, _InputIterator1 __last1,
     ^~~~~~~~~
/usr/include/c++/7/bits/stl_algo.h:4332:5: note:   template argument deduction/substitution failed:
lexer.cc:265:84: note:   candidate expects 5 arguments, 4 provided
     std::transform(ciSuffix.begin(), ciSuffix.end(), ciSuffix.begin(), std::toupper);
                                                                                    ^
lexer.cc: In member function ‘std::tuple<std::shared_ptr<SyntaxToken>, std::shared_ptr<StrayToken>, bool, bool> Lexer::ReadTokenOnce()’:
lexer.cc:806:21: warning: unused variable ‘base’ [-Wunused-variable]
         const char *base{ &l->Source->Lines[line][column] };
                     ^~~~
Makefile:26: recipe for target 'compile' failed
make: *** [compile] Error 1
```

**Comment by theKidOfArcrania** [comment]
okay I fixed the compile errors (from this patch):
```patch
diff --git a/compiler/Makefile b/compiler/Makefile
index 7dc9bbf..c67db94 100644
--- a/compiler/Makefile
+++ b/compiler/Makefile
@@ -1,6 +1,6 @@
 TARGET         := compile
 
-CXXFLAGS       := -g -std=gnu++14 -Wall -Wextra
+CXXFLAGS       := -g -std=gnu++1z -pedantic -Wall -Wextra
 
 HHFILES        := \
        lexer.hh \
diff --git a/compiler/lexer.cc b/compiler/lexer.cc
index 654fdcf..819c012 100644
--- a/compiler/lexer.cc
+++ b/compiler/lexer.cc
@@ -1,11 +1,11 @@
-#define _CRT_SECURE_NO_WARNINGS
+//#define _CRT_SECURE_NO_WARNINGS
 #include "lexer.hh"
 #include "logger.hh"
-#include <assert.h>
-#include <stdint.h>
-#include <stdio.h>
-#include <stdlib.h>
-#include <string.h>
+#include <cassert>
+#include <cstdint>
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
 #include <algorithm>
 
 #define LM_DEFAULT                   0
@@ -39,8 +39,9 @@ Rc<SyntaxToken> Lexer::ReadTokenDirect() {
         l->CurrentToken = token;
 
         if (asStrayToken) {
+            auto range = l->CurrentToken->GetLexemeRange();
             LogAtRange(
-                &l->CurrentToken->GetLexemeRange(),
+                &range,
                 LL_ERROR,
                 "stray '%c' in program",
                 asStrayToken->GetOffendingChar()
@@ -262,7 +263,7 @@ std::string Lexer::ReadSuffix() {
 void Lexer::SkipIntSuffixes(const Rc<SyntaxToken> t) {
     std::string suffix{ ReadSuffix() };
     std::string ciSuffix{ suffix };
-    std::transform(ciSuffix.begin(), ciSuffix.end(), ciSuffix.begin(), std::toupper);
+    std::transform(ciSuffix.begin(), ciSuffix.end(), ciSuffix.begin(), toupper);
 
     if (ciSuffix.length() != 0 &&
         ciSuffix != "UL" &&
```
Now i get linkage errors now:
```
[...]
/tmp/ccn13Yv7.o: In function `void __gnu_cxx::new_allocator<WarningDirectiveKw>::construct<WarningDirectiveKw>(WarningDirectiveKw*)':
/usr/include/c++/7/ext/new_allocator.h:136: undefined reference to `WarningDirectiveKw::WarningDirectiveKw()'
/tmp/ccn13Yv7.o: In function `void __gnu_cxx::new_allocator<ConstKeyword>::construct<ConstKeyword>(ConstKeyword*)':
/usr/include/c++/7/ext/new_allocator.h:136: undefined reference to `ConstKeyword::ConstKeyword()'
/tmp/ccn13Yv7.o: In function `void __gnu_cxx::new_allocator<ExternKeyword>::construct<ExternKeyword>(ExternKeyword*)':
/usr/include/c++/7/ext/new_allocator.h:136: undefined reference to `ExternKeyword::ExternKeyword()'
/tmp/ccn13Yv7.o: In function `void __gnu_cxx::new_allocator<StaticKeyword>::construct<StaticKeyword>(StaticKeyword*)':
/usr/include/c++/7/ext/new_allocator.h:136: undefined reference to `StaticKeyword::StaticKeyword()'
/tmp/ccn13Yv7.o: In function `void __gnu_cxx::new_allocator<AutoKeyword>::construct<AutoKeyword>(AutoKeyword*)':
/usr/include/c++/7/ext/new_allocator.h:136: undefined reference to `AutoKeyword::AutoKeyword()'
/tmp/ccn13Yv7.o: In function `void __gnu_cxx::new_allocator<VolatileKeyword>::construct<VolatileKeyword>(VolatileKeyword*)':
/usr/include/c++/7/ext/new_allocator.h:136: undefined reference to `VolatileKeyword::VolatileKeyword()'
/tmp/ccn13Yv7.o: In function `void __gnu_cxx::new_allocator<UnsignedKeyword>::construct<UnsignedKeyword>(UnsignedKeyword*)':
/usr/include/c++/7/ext/new_allocator.h:136: undefined reference to `UnsignedKeyword::UnsignedKeyword()'
/tmp/ccn13Yv7.o: In function `void __gnu_cxx::new_allocator<SignedKeyword>::construct<SignedKeyword>(SignedKeyword*)':
/usr/include/c++/7/ext/new_allocator.h:136: undefined reference to `SignedKeyword::SignedKeyword()'
/tmp/ccn13Yv7.o: In function `void __gnu_cxx::new_allocator<VoidKeyword>::construct<VoidKeyword>(VoidKeyword*)':
/usr/include/c++/7/ext/new_allocator.h:136: undefined reference to `VoidKeyword::VoidKeyword()'
/tmp/ccn13Yv7.o: In function `void __gnu_cxx::new_allocator<CharKeyword>::construct<CharKeyword>(CharKeyword*)':
/usr/include/c++/7/ext/new_allocator.h:136: undefined reference to `CharKeyword::CharKeyword()'
/tmp/ccn13Yv7.o: In function `void __gnu_cxx::new_allocator<ShortKeyword>::construct<ShortKeyword>(ShortKeyword*)':
[...]

**Comment by josh-127** [comment]
Oh, I was porting the compiler over to Windows (from issue #9 ), but I forgot to update the Makefiles. I'll merge these changes as soon as possible.

**Comment by josh-127** [comment]
Unfortunately, Ubuntu 16.04's package repository doesn't have the latest version of GCC, so I'm stuck with version 5.4.0. Gonna use a PPA.

**Comment by josh-127** [comment]
Commit 01b0bd1:

```
Fixes build issues on Ubuntu 16.04 LTS and above. If you're using
Ubuntu 16.04 LTS, then you must install GCC 8.x from a Personal
Package Archive (PPA). You can find one in Launchpad.net, or follow
these instructions:

    In your terminal:
        1. sudo apt-add-repository ppa:ubuntu-toolchain-r/test
        2. sudo apt-get update
        3. sudo apt-get install gcc-8 g++-8

----------------------------------------------------------------

For issue #10
```

**Comment by josh-127** [comment]
Forgot to reference this issue in another commit, so here it is:
Updated README instructions in commit 5876e41e4ebcb5d0bab101d1713b6751b8112f20 .
